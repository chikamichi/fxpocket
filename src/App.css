@-webkit-keyframes sk-scaleout {
  0% { -webkit-transform: scale(0) }
  100% {
    -webkit-transform: scale(1.0);
    opacity: 0;
  }
}

@keyframes sk-scaleout {
  0% {
    -webkit-transform: scale(0);
    transform: scale(0);
  } 100% {
    -webkit-transform: scale(1.0);
    transform: scale(1.0);
    opacity: 0;
  }
}

body {
  background: cornsilk;
}

#root {
  display: flex;
  justify-content: center;
}

.fxp {
  text-align: center;
}

.fxp-title {
  margin-bottom: 0;
  color: brown;
}

.fxp-motto {
  margin-top: 0;
  font-size: 1em;
  color: slategray;
}

/* fxPocket App => fxp- namespace */
.fxp-app {
  display: flex;
  margin: 2em 0;
}

/* Borrowed from http://tobiasahlin.com/spinkit/ */
.fxp-loading {
  width: 40px;
  height: 40px;
  margin: 100px auto;
  background-color: #333;
  border-radius: 100%;
  -webkit-animation: sk-scaleout 1.0s infinite ease-in-out;
  animation: sk-scaleout 1.0s infinite ease-in-out;
}

.fxp-currency {
  display: flex;
  flex-grow: 1;
  flex-direction: column;
  justify-content: flex-end;
  margin: 10px;
  box-shadow: 2px 2px lightgrey;
  padding: 4px;
  border-radius: 4px;
  background: wheat;
}

.fxp-currency__list {
  color: lightgrey;
}

.fxp-currency__result {
  padding: 10px;
  background: white;
  border: 1px solid lightgrey;
}

.fxp-currency__amount {
  border: none;
  background: transparent;
}

.fxp-currency__label {
  padding-left: 0.5em;
  vertical-align: middle;
}

.fxp-footer-links {
  list-style: none;
  font-size: 0.8em;
}

.fxp-footer-links .fxp-u-list__item {
  display: inline;
  margin-left: 1em;
}

.fxp-footer-links .fxp-u-list__item:first-child {
  margin-left: 0;
}

.fxp-footer-link,
.fxp-footer-link:visited {
  color: grey;
  text-decoration: none;
}

.fxp-footer-link:hover {
  text-decoration: underline;
}
