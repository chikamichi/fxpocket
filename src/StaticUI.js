import React, { Component } from 'react'

import { APP_TITLE, APP_MOTTO } from './utils'

/*
 * StaticUI â€” A presentational wrapper to prettify the core app.
 */
export default (WrappedComponent) => {
  return class StaticUI extends Component {
    render() {
      return (
        <div className="fxp">
          <h1 className="fxp-title">{APP_TITLE}</h1>
          <h2 className="fxp-motto">{APP_MOTTO}</h2>
          <WrappedComponent {...this.props} />
        </div>
      )
    }
  }
}
