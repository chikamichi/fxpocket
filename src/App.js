import React, { Component } from 'react'
import 'whatwg-fetch'

import Currency from './Currency'
import { round } from './utils'
import Fixer from './fixer'
import './App.css'

const INITIAL_STATE = {
  // Keeps track of app's startup process.
  init: false,
  // A list of currency widgets to display at startup.
  widgets: ['EUR', 'USD'],
  // Keeps track of user intents.
  base: {
    amount: undefined,
    currency: undefined
  }
}

class App extends Component {
  constructor(props) {
    super(props)
    this.state = INITIAL_STATE
  }

  onAmountEdited(payload) {
    this.setState({base: payload})
  }

  onCurrencyEdited(payload) {
    const idx = payload.uuid
    const newWidgetsState = [
      ...this.state.widgets.slice(0, idx),
      payload.currency,
      ...this.state.widgets.slice(idx+1)
    ]
    const newBaseState = {
      amount: payload.amount,
      currency: payload.currency
    }
    this.setState({
      widgets: newWidgetsState,
      base: newBaseState
    })
  }

  convertAmounts() {
    const base = this.state.base
    if (!base.amount || !base.currency) return {}
    return this.fixer.currencies.reduce((acc, currency) => {
      let amount
      if (base.currency === currency)
        amount = base.amount
      else
        amount = base.amount * this.fixer.rates[base.currency][currency]
      acc[currency] = round(amount, 2)
      return acc
    }, {})
  }

  componentDidMount() {
    Fixer.connect().then(fixer => {
      this.fixer = fixer
      this.setState({init: true})
    })
  }

  render() {
    if (!this.state.init) return <p>loading app</p>
    const currencies = this.fixer.currencies
    const cb = {
      onAmountEdited: this.onAmountEdited.bind(this),
      onCurrencyEdited: this.onCurrencyEdited.bind(this)
    }
    const amounts = this.convertAmounts()
    const widgets = this.state.widgets.map((currency, i) => {
      return <Currency
        key={i} // key is simply the position in this.state.widgets
        uuid={i}
        type='quote' // TODO: remove this type prop
        amount={amounts[currency]}
        currency={currency}
        currencies={currencies}
        onAmountEdited={cb.onAmountEdited}
        onCurrencyEdited={cb.onCurrencyEdited} />
    })
    return (
      <div className="fxp">
        {widgets}
      </div>
    )
  }
}

export default App
