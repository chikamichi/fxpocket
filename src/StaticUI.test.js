import React from 'react'
import { shallow } from 'enzyme'

import App from './App';
import StaticUI from './StaticUI'

const PrettyApp = StaticUI(App)

describe('Wrapping, static UI', () => {
  describe('layout', () => {
    var wrapper

    beforeEach(async () => {
      const PrettyApp = StaticUI(App)
      wrapper = await shallow(<PrettyApp />)
      wrapper.update()
    })

    it('looks like the registered snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })

})
