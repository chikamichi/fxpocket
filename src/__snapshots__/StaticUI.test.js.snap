// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wrapping, static UI layout looks like the registered snapshot 1`] = `
<div
  className="fxp"
>
  <header>
    <h1
      className="fxp-title"
    >
      fxPocket
    </h1>
    <h2
      className="fxp-motto"
    >
      Currency conversion made easy.
    </h2>
  </header>
  <App />
  <footer>
    <ul
      className="fxp-u-list  fxp-footer-links"
    >
      <li
        className="fxp-u-list__item"
      >
        <a
          className="fxp-footer-link"
          href="https://github.com/chikamichi/fxpocket/"
        >
          source
        </a>
      </li>
      <li
        className="fxp-u-list__item"
      >
        <a
          className="fxp-footer-link"
          href="https://en.wikipedia.org/wiki/Exchange_rate"
        >
          about exchange rates
        </a>
      </li>
    </ul>
  </footer>
</div>
`;
